src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs

```diff
@@ -180,11 +190,44 @@ public Task StopAsync()
             catch (Exception ex)
             {
                 Log.UnexpectedExceptionInStop(Logger, ConnectionId, ex);
-            }
-            
+            }
+
             return Task.CompletedTask;
         }
 
+        public async Task CloseAsync()
+        {
+            if (Interlocked.CompareExchange(ref _closeStage, ClosingSendingFin1, ClosingDefault) != ClosingDefault)
+            {
+                // CloseAsync is in progress or done.
+                return;
+            }
+
+            while (Interlocked.CompareExchange(ref _closeStage, ClosingWaitingApplicationTasks, ClosingReceivedFinAck1) != ClosingReceivedFinAck1)
+            {
+                await Task.Delay(TimeSpan.FromMilliseconds(50));
+            }
+
+            if (clientOutgoingTasks.Values.Count > 0)
+            {
+                await Task.WhenAll(clientOutgoingTasks.Values);
```

Comment:
>clientOutgoingTasks [](start = 35, length = 19)

what if it never ends?

Reply:
This is the endless version.
There is another method with a timeout `CloseAsync(TimeSpan timeout)` implemented by this method.